core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -55,16 +56,94 @@ namespace kagome::blockchain {
+
+    // Find the least and best leaf
+    primitives::BlockInfo least_leaf(
+        std::numeric_limits<primitives::BlockNumber>::max(), {});
+    primitives::BlockInfo best_leaf(
+        std::numeric_limits<primitives::BlockNumber>::min(), {});
+    for (auto hash : block_tree_leaves) {
+      OUTCOME_TRY(number, header_repo->getNumberById(hash));
+      if (number <= least_leaf.number) {
+        least_leaf = {number, hash};
+      }
+      if (number >= best_leaf.number) {
+        best_leaf = {number, hash};
+      }
+    }
+
     std::optional<consensus::EpochNumber> curr_epoch_number;
+
+    // First, look up slot number of block number 1
+    OUTCOME_TRY(first_block_header_exists, storage->hasBlockHeader(1));
+    if (first_block_header_exists) {
+      OUTCOME_TRY(first_block_header, storage->getBlockHeader(1));
+      auto babe_digest_res = consensus::getBabeDigests(first_block_header);
+      BOOST_ASSERT_MSG(babe_digest_res.has_value(),
+                       "Any non genesis block must be contain babe digest");
+      auto first_slot_number = babe_digest_res.value().second.slot_number;
+
+      // Second, look up slot number of best block
+      auto &best_block_hash = best_leaf.hash;
+      auto best_block_header_res = storage->getBlockHeader(best_block_hash);
+      BOOST_ASSERT_MSG(best_block_header_res.has_value(),
+                       "Best block must be known whenever");
+      const auto &best_block_header = best_block_header_res.value();
+      babe_digest_res = consensus::getBabeDigests(best_block_header);
+      BOOST_ASSERT_MSG(babe_digest_res.has_value(),
+                       "Any non genesis block must be contain babe digest");
+      auto last_slot_number = babe_digest_res.value().second.slot_number;
+
+      BOOST_ASSERT_MSG(
+          last_slot_number >= first_slot_number,
+          "Non genesis slot must not be less then slot of block number 1");
+
+      // Now we have all to calculate epoch number
+      auto epoch_number = (last_slot_number - first_slot_number)
+                          / babe_configuration->epoch_length;
+      consensus::EpochDescriptor epoch{
+          .epoch_number = epoch_number,
+          .start_slot = first_slot_number
+                        + epoch_number * babe_configuration->epoch_length};
+
+      babe_util->syncEpoch(epoch);
+
+      curr_epoch_number.emplace(epoch_number);
+    }
+
+    primitives::BlockInfo last_finalized_block_info;
+
+    // Backward search of finalized block
+    for (auto block_info = least_leaf;;) {
+      if (block_info.number == 0) {
+        last_finalized_block_info = block_info;
+        break;
+      }
+      auto j_res = storage->getJustification(block_info.hash);
+      if (j_res.has_value()) {
+        last_finalized_block_info = block_info;
+        break;
+      }
+      if (j_res
+          != outcome::failure(
+              BlockStorageError::JUSTIFICATION_DOES_NOT_EXIST)) {
+        return j_res.as_failure();
```

Comment:
you don't need explicit conversions via as_failure.
When j_res does not have a value it could be returned as-is or via j_res.error().

There are no reasons to use as_failure().

Reply:
Not agree. j_res differs of return type (**outcome::result<std::shared_ptr<BlockTreeImpl>>** vs. **outcome::result<primitives::Justification>**) and must be cast explicitly.